### ISSUE TITLE
Allow setting default account email per organization

### ISSUE THREAD

**Issue by willnorris** [todogroup/gh-issues#32]
You can set multiple email addresses on your GitHub account, and can [route notifications on a per-org basis](https://help.github.com/articles/managing-notification-emails-for-organizations/).  However, when you create a commit through github.com (either through the web editor, or merge commits generated by the green merge button), the commit is always authored by the user's default email address (unless of course they have email privacy enabled).  

I would like to be able to set the default email so that any changes made in a github.com/google repo always uses my @google.com email.  (This is particularly important for us because we want employees to always use their work email address for Google repos)

**Comment by bpapa** [comment]
Would love this because it's broken a key workflow in our org. We're using Xcode server, and the "Big Screen" web view for CI Reports makes it look like we have more team members than we actually do, because some of our commits are from our local machines and some are merges done using the Web UI.

**Comment by adam-arold** [comment]
Bump. Is there some progress with this? I dont want to have to maintain more than one GH profile if I can help it but without this I have to. Is there a solution for this?

**Comment by clarkbw** [comment]
Hold tight, something is coming

**Comment by damienjbyrne** [comment]
Any progress on this?  This is a feature I'd really like to have.

**Comment by benbalter** [comment]
> I would like to be able to set the default email so that any changes made in a github.com/google repo always uses my @google.com email. 

Please excuse the lack of update while @clarkbw is away, but just discovered this issue, @willnorris and wanted to point you to https://blog.github.com/changelog/2018-05-03-choose-commit-email/.

As of May, you can choose the email used for web-based commits on the blob editor. It will default to the last email used to commit to the repo, and if you haven't committed before, your org-specific email. 

Due to UI complexity and frequency of use, we scoped the feature to just blob edit commits, so merge commits are unaffected at this time and continue to be authored by your primary email.

**Comment by adam-arold** [comment]
I think for most of us the main point of this feature is to *have a separate email for merging pull requests*. This is literally true for everyone who has a GitHub account for personal use and wants to use it at the workplace which also uses GitHub.

**Comment by jackowayed** [comment]
I just wanted to chime in and say that I've built a tool for my organization that is creating and merging PR's via the API, and we have the same issue. It's exacerbating our existing problem with people committing with personal emails from manually merging PR's.

So I hope that when someone takes a look at doing this for PR's, they keep the API in mind. (Or maybe do the API sooner because there isn't the same UI complexity? 😃)

**Comment by joaoe** [comment]
+1
This is a serious deal for me.
My organization uses github and every commit done on the web interface, typically merging PRs uses my personal e-mail and no the e-mail of my employer, which is quite a serious problem.

**Comment by dipla65** [comment]
Hi,
How can i change the content in the email on a timebank application? Is anyone able to help.
Many thanks

**Comment by adam-arold** [comment]
@clarkbw can you elaborate on that something which is coming?

**Comment by dipla65** [comment]
HI Adam,

Dipak Mistry

IT Support

Semcon Product Information UK Ltd

Semcon House

Edgehill Drive

WARWICK,

CV34 6NH

Mobile: +44 (0) 7789 488051

Email:    dipak.mistry@semcon.com



www.semcon.com

Registered Office: Semcon Product Information UK Ltd, Semcon House, Edgehill Drive, Warwick, Warwickshire CV34 6NH

Registered Office in England No. 06132347

Please consider the environment before printing this email.

**Comment by dipla65** [comment]
Hi Adam,
My issue is I have an application in Github which someone created.

This sends an email with automated response from our network.  It is for a timebank application.

I can send you the application to give  you an idea.  It is about 1mb.

Dipak Mistry

IT Support

Semcon Product Information UK Ltd

Semcon House

Edgehill Drive

WARWICK,

CV34 6NH

Mobile: +44 (0) 7789 488051

Email:    dipak.mistry@semcon.com



www.semcon.com

Registered Office: Semcon Product Information UK Ltd, Semcon House, Edgehill Drive, Warwick, Warwickshire CV34 6NH

Registered Office in England No. 06132347

Please consider the environment before printing this email.

**Comment by benbalter** [comment]
@adam-arold I _believe_ @clarkbw may have been referring to https://blog.github.com/changelog/2018-05-03-choose-commit-email/, which solved (I think) the problem @willnorris described initially, at least in terms of web edits. 

Building on that, https://github.blog/changelog/2019-01-23-limit-organization-email-notifications-to-verified-domains/ further segments personal and corporate email. 

The last remaining step, at least in terms of this thread, would be the ability to choose your merge commit email. While I can't promise a specific date, it's one of our top feature requests, and I'm optimistic it should be implemented in the coming months.

**Comment by adam-arold** [comment]
Nice, thanks!

**Comment by ehartung** [comment]
@benbalter The feature you mentioned is about setting the commit email address per account, but the issue is about setting a default email address per organisation. This is especially important with GitHub Enterprise Cloud when you can use your account for private development (personal email) and professional work (company email).

**Comment by benbalter** [comment]
> setting a default email address per organisation.

@ehartung are you able to elaborate? Are you concerned about personal emails being used for notifications? For web commits? For commits pushed via command line? Something else?

**Comment by ehartung** [comment]
@benbalter it's about having my personal email address for merge (web) commits (after PR) for private projects and my company email address for merge commits for business projects (in the GitHub-Org of the company I work for)

**Comment by ehartung** [comment]
... when I talk about "business projects" I mean, non-open-source projects in private repositories (GitHub Enterprise Cloud)

**Comment by benbalter** [comment]
>it's about having my personal email address for merge (web) commits (after PR) for private projects and my company email address for merge commits for business projects (in the GitHub-Org of the company I work for)

@ehartung thanks for clarifying. Please see [my comment above](https://github.com/todogroup/gh-issues/issues/32#issuecomment-461043653):

> The last remaining step, at least in terms of this thread, would be the ability to choose your merge commit email. While I can't promise a specific date, it's one of our top feature requests, and I'm optimistic it should be implemented in the coming months.

As a first pass, we'll allow users to choose any verified email to use for merge commits (as we do for web edits), defaulting to the last email used in the repository, or the organization-specific notification email, if one is set. For most cases, this would _default_ to the corporate email in business projects and their person email in personal projects. The next step would be to allow organization admins to enforce that the corporate email is used for merge commits, which is on our roadmap as a subsequent follow up after the ability to choose the merge commit email.

**Comment by ehartung** [comment]
@benbalter Thanks! Having the ability to choose the email address for merge commits would be great.

**Comment by timoreimann** [comment]
I believe https://github.com/isaacs/github/issues/95 is asking for the very same feature.

**Comment by ishepherd** [comment]
Ping. Please can you update us on the ETA, @benbalter ?

**Comment by numbsafari** [comment]
Will the ability to specify a default account per-organization be something that only applies to the Enterprise cloud, meaning that it's yet another reason not to host an open-source project on github.com, or will it be something that is available for regular github.com organizations?

**Comment by benbalter** [comment]
> Please can you update us on the ETA

I can't promise an exact timeline for general availability, but for those that are members of the Maintainer's Early Access Program, you should expect access to a private beta within the coming week.

> Will the ability to specify a default account per-organization be something that only applies to the Enterprise cloud

Per [my comment above](https://github.com/todogroup/gh-issues/issues/32#issuecomment-461451919), as we do with commits from web edits today, if you have multiple verified emails, you will be presented with the option to choose your merge commit email (on all repos). We'll try to anticipate the email you'd like to use (the default in the dropdown, which you can change), based on your prior commits to the repo, or if this is your first commit, your notification email for the organization, if one is set. This will similarly be available on all repositories. 

Building on that, down the line, we might offer organization admins the ability to _restrict_ which emails can be used for merge commits (e.g., requiring an email in the organization's verified domain be used, and only listing those email in the dropdown). That's a bit further down the line, as we wanted to get the ability to choose your merge commit email out first and in your hands, along with several other improvements, and haven't yet decided if the ability to restrict merge commit emails will be limited to organizations on the Enterprise Cloud plan.

**Comment by benbalter** [comment]
> for those that are members of the Maintainer's Early Access Program, you should expect access to a private beta within the coming week.

If you are a member of the maintainer's early access program, please see https://github.com/maintainers/early-access-feedback/issues/235.

**Comment by benbalter** [comment]
Thank you all for your thoughtful feedback. You can now choose the author email for commits created when merging pull requests via the web.  For more information, see https://github.blog/changelog/2019-04-24-choose-your-merge-commit-email/.

**Comment by LilyMGoh** [comment]
@benbalter Is there a way to set the organization default email so that I don't have to select from the dropdown every time? Thanks.

**Comment by benbalter** [comment]
> Is there a way to set the organization default email so that I don't have to select from the dropdown every time? 

@LilyMGoh the drop down will default to the last email you used to commit to the repo, defaulting to your organization notification email if this is your first commit, meaning ideally, it should default to the email you'd like to use for the repo, or will after your first use, without needing to manage an additional setting.

**Comment by LilyMGoh** [comment]
@benbalter It's a bug then.

**Comment by fedepires** [comment]
> @benbalter It's a bug then.

I agree. Would be nice to be able to set the organization's default email for every repo under such org.

**Comment by tylerb** [comment]
The drop down does not remember the last setting I used for the organization. It always resets to my default github email account.

**Comment by benbalter** [comment]
@tylerb are you able to share the steps you take to see that? I just tried to reproduce it and wasn't able to under the following conditions:

1. As a user with two verified emails, A and B
2. Set primary email to email A
3. In an org owned repo, edit a file via the web editor
4. See that it defaults to email A
5. Select to commit with email B and commit
4. Edit the file again
5. See that it defaults to email B
6. Create a pull request
7. See that the merge box defaults to email B

Are you doing something differently? Would love to better understand what's going on so that we can fix it. Thanks!

**Comment by tylerb** [comment]
Certainly:

1. As a user with two verified emails, A and B
2. Set primary email to email A
3. Create a local branch
4. Commit some code
5. Push to a new, remote branch
6. Open a PR
7. Click "Squash and Merge"
8. See that the dropdown shows email A
9. Choose email B
10. Complete the merge
11. Run steps 3 through 7
12. See that the dropdown *still* shows email A

**Comment by benbalter** [comment]
> Commit some code

@tylerb Is it possible that the local commits are authored with email A? If so, would you expect the squash and merge button to change the author email by default?

**Comment by tylerb** [comment]
@benbalter That may be true, yes. However, the local commits in the PRs I'm working on are created by various members of my team, then I am the one who does the squash and merge from the feature branch into the main branch.

**Comment by mattt** [comment]
@benbalter Thanks so much for the ability to configure email addresses for merge commits — this is really helpful! Would it be possible to extend this functionality when resolving merge conflicts on GitHub.com? Currently, it doesn't look like there's a way to select a different email from the primary / default address associated with your account.

**Comment by benbalter** [comment]
> Currently, it doesn't look like there's a way to select a different email from the primary / default address associated with your account.

That is correct. Due to limitations in the user interface, for this iteration we chose to _display_ the email that will be used to resolve merge conflicts via the web UI, but hope to make that configurable in a subsequent iteration. Thanks for the feedback!

**Comment by benbalter** [comment]
@tylerb dug into this a bit more and wanted to share two scenarios where I believe the drop down may not reflect the last email used _in the drop down_:

1. When after merging, commits are pushed to any branch using a different email.
2. When squash merging another user's pull request

The reason for both is that we look for the most recent commit, on any branch, that you _authored_. If you've authored other commits since the last merge via the web interface using another email, those commits will override the preference. Similarly, if you squash merge another user's pull request, the actor is set as the committer and the PR author is set as the commit author (for proper attribution), in which case the email you selected via the dropdown for that squash merge (of another user's pull request) would not be preserved and it would default to your most recent authored commit or your per-organization notification email, if set.

**Comment by tylerb** [comment]
@benbalter perfect! That really helps me understand why it's happening so I can get the behavior I want. Thanks for your help and time on this!

**Comment by waldyrious** [comment]
> Would it be possible to extend this functionality when resolving merge conflicts on GitHub.com? Currently, it doesn't look like there's a way to select a different email from the primary / default address associated with your account.

> That is correct. Due to limitations in the user interface, for this iteration we chose to display the email that will be used to resolve merge conflicts via the web UI, but hope to make that configurable in a subsequent iteration. Thanks for the feedback!

For the record, it seems to also be impossible to change the committer email when rebasing a PR.

I just had this problem when merging PRs via rebase in an organization repo, and the only way to make the committer email reflect the email associated with the organization was to change the primary email of my account, merge the PR, and then change it back (hopefully that preference will now be preserved for future rebase-based merges). Looking forward to have the ability to select author/committer email in the UI in more cases.

**Comment by vividn** [comment]
For your local repos make sure to associate your work email with the repo so that commits you push remotely will have the correct associated email. Then the dropdown will be correctly associated with your work when doing a PR.
`git config user.email myworkemail@workplace.com`

**Comment by fleetofthemalden** [comment]
I am unable to see a dropdown. Just text saying it will be associated with my primary email even though all my commits thus far have been for my org email.

**Comment by vividn** [comment]
@fleetofthemalden, are your other emails added and verified in your settings on github.com?
github > settings > emails
other than that :man_shrugging:

**Comment by fleetofthemalden** [comment]
Yep. Added, verified, all commits are from the alternate (work) email, but the web UI is informing me that the merge commit will be made with my primary (personal) email, and no dropdown is available.

> @fleetofthemalden, are your other emails added and verified in your settings on github.com?
> github > settings > emails
> other than that 🤷‍♂

**Comment by GMNGeoffrey** [comment]
A similar issue I'm running into is that when pulling down GitHub PRs via the API. The GitHub generated merge commit for a PR always uses the primary email address. That means that even if I have a PR with a single commit that uses one email, the GitHub generated merge commit uses another email

```
$ git log -1 origin/pull/142/head
commit de7bae8fbb961a6dce3d8aa9a648aaa935ddf773 (origin/test-pr-1, origin/pull/142/head, test-pr-1)
Author: Geoffrey Martin-Noble <gcmn@google.com>
Date:   Wed Nov 13 14:17:57 2019 -0800

    Test PR

$ git log -1 origin/pull/142/merge
commit 62c3308ca3479bba702cc68f2e56737a85de905b (origin/pull/142/merge)
Merge: 4004638 de7bae8
Author: Geoffrey Martin-Noble <GMNGeoffrey@gmail.com>
Date:   Thu Nov 14 15:48:34 2019 -0500

    Merge de7bae8fbb961a6dce3d8aa9a648aaa935ddf773 into 4004638b935d5dd49aa27507f3371110bd3c174c
```

More generally, it seems like the approach here has been somewhat piecemeal: "allow setting the committer email in these places", whereas what I (and AFAICT most people on this thread) want is "make all interactions with this organization/repository use this email". I'm having trouble imagining a use case where I want to be able to mix and match which email is associated with a repository or organization.

**Comment by jlevon** [comment]
There's no API field to allow us to set this when doing a squash merge, as far as I can tell. Is there an issue open on that?

**Comment by karora** [comment]
The issue for me is when other people in ${private organisation} commit (or whatever) to ${private organisation repository}, and the notifications of them having done that get sent to my personal e-mail address.  I want all communication in relation to provide company repositories to go to my private company e-mail, whether I caused the change or not.

**Comment by GMNGeoffrey** [comment]
> The issue for me is when other people in ${private organisation} commit (or whatever) to ${private organisation repository}, and the notifications of them having done that get sent to my personal e-mail address. I want all communication in relation to provide company repositories to go to my private company e-mail, whether I caused the change or not.

That is something you can actually set at https://github.com/settings/notifications. You can set notification email per organization.

**Comment by benbalter** [comment]
> That is something you can actually set at https://github.com/settings/notifications. You can set notification email per organization.

Organization owners can also force the setting for all organization members as well: https://help.github.com/en/github/setting-up-and-managing-organizations-and-teams/restricting-email-notifications-to-an-approved-domain

**Comment by karora** [comment]
@benbalter Thanks!  My searching had failed to find that (and found this in preference), and I guess I'd been expecting it to be on the Organisations page of the settings.

**Comment by jeffmcaffer** [comment]
Can we close this now?

**Comment by GMNGeoffrey** [comment]
> Can we close this now?

The question karora just asked is largely unrelated to the rest of the thread. The core issue remains outstanding

**Comment by GMNGeoffrey** [comment]
Yet another place where this comes up. When making suggested edits it uses the default email address to populate the "Co-authored-by" tag

**Comment by rfgamaral** [comment]
> 
> 
> Yet another place where this comes up. When making suggested edits it uses the default email address to populate the "Co-authored-by" tag

Came here looking for this option...

Why can't we have a single place to configure the default e-mail address per organization for **everything** related to that organization?

**Comment by djfurie** [comment]
Came here looking for a solution, hoping that surely this was solved sometime in the past 4 years?

I see I can get notifications per-organization, but is it still true that I cannot set my preferred email for Pull Request Merge Commits via the web interface on a per organization basis?

**Comment by joshkpeterson** [comment]
Came here via a google search for setting email per-organization, and https://github.com/settings/notifications was what I was looking for. Might I also suggest adding a link to these notifications settings on Settings > Emails? 

Besides being related, it also fits with the flow of being onboarded to a new gig and adding your new email address with them to your github account. The user is already right there, if that makes sense.

**Comment by GMNGeoffrey** [comment]
Found another case. When using "squash and merge" on someone's PR it will sometimes use a different email address for the author than the email address used for every commit in the PR. I'm guessing it uses whatever they have set as their primary email? Or maybe they missed one of the many dropdowns. Please fix this by adding per-organization email address association as originally described in this report 4 years ago.

**Comment by ricky-wong** [comment]
Yup basically isaacs/github#1368. My org is having quite a bit of trouble due to this.

Did not realize it was just hidden, maybe an extension to flip the `display: none` here would help. https://github.com/isaacs/github/issues/1690#issuecomment-583309913

**Comment by GMNGeoffrey** [comment]
> Did not realize it was just hidden, maybe an extension to flip the `display: none` here would help. [isaacs/github#1690 (comment)](https://github.com/isaacs/github/issues/1690#issuecomment-583309913)

I don't think that would fix it. That's a dropdown I've seen before which allows me as the person doing the "squash and merge" to set which email I want to use (as the committer, despite the CSS class name). It came about in response to this bug. It doesn't help in the case of GitHub changing the author of the commit.

**Comment by ricky-wong** [comment]
> > Did not realize it was just hidden, maybe an extension to flip the `display: none` here would help. [isaacs/github#1690 (comment)](https://github.com/isaacs/github/issues/1690#issuecomment-583309913)
> 
> I don't think that would fix it. That's a dropdown I've seen before which allows me as the person doing the "squash and merge" to set which email I want to use (as the committer, despite the CSS class name). It came about in response to this bug. It doesn't help in the case of GitHub changing the author of the commit.

I just used a test repo to verify that you're correct. Oops.
Despite selecting a different email, it still uses the initial/primary email of the squasher/committer (in our org, they happen to be the same most of the time)

**Comment by fredthomsen** [comment]
> 
> 
> Found another case. When using "squash and merge" on someone's PR it will sometimes use a different email address for the author than the email address used for every commit in the PR. I'm guessing it uses whatever they have set as their primary email? Or maybe they missed one of the many dropdowns. Please fix this by adding per-organization email address association as originally described in this report 4 years ago.

I just got bit by this as well.   I do open source contributions as part of work, and obviously my company would like to see the commit authored by me using a company email address rather than my personal.

**Comment by GMNGeoffrey** [comment]
The only "solution" I've come across is to set my work email as my default email. Nothing else works fully

**Comment by admxxi** [comment]
> The only "solution" I've come across is to set my work email as my default email. Nothing else works fully

Yeap, did the same here 😢

**Comment by sulfasTor** [comment]
> > The only "solution" I've come across is to set my work email as my default email. Nothing else works fully
> 
> Yeap, did the same here 😢

+1 Hoping this feature is added soon.

**Comment by ulfryk** [comment]
+1

**Comment by zomars** [comment]
Maybe we can get this discussion get some traction in here: https://github.com/github/feedback/discussions/4168

**Comment by bidyashish** [comment]
If work email is verified, then notification emails from work organisation can be routed to work email, without changing the primary email

Email change dropdown are under

Settings > notifications > Custom routing per Organisation (bottom of page)

<img width="788" alt="Screenshot 2021-11-06 at 1 48 36 AM" src="https://user-images.githubusercontent.com/14001930/140555676-4122bd8d-fc6f-46f1-87e1-bb840420359f.png">

**Comment by garretmh** [comment]
@bidyashish That is mentioned right away in the issue description and is not what this issue is about.

**Comment by johnpitchko** [comment]
+1

**Comment by Bullsized** [comment]
Hopefully I am helping someone looking for this in 2022:
[Settings -> Notifications](https://github.com/settings/notifications), scroll down to bottom, there should be **Custom routing** section in which you can reroute organization mails to one of your verified emails.

**Comment by bencooper222** [comment]
That doesn't quite solve the problem. For example, if I make a quick change in the GH web UI, if I'm on a company repo, it should default to using my company email address.

**Comment by eKoopmans** [comment]
Still looking for a fix for this. Worth noting, for a while there was a (great!) feature where the commit author could be selected from a dropdown when merging:

- the feature was added April 24, 2019 by @benbalter:
	https://github.com/todogroup/gh-issues/issues/32#issuecomment-486382937
- the feature was removed again sometime before Feb 10, 2021 (possibly before Feb 7, 2020):
	https://github.com/todogroup/gh-issues/issues/32#issuecomment-777061396
	https://github.com/isaacs/github/issues/1690#issuecomment-583309913
- the dropdown is still there (hidden), but using it has no effect:
	https://github.com/todogroup/gh-issues/issues/32#issuecomment-777077553
	https://github.com/isaacs/github/issues/1740#issuecomment-651710960
- the dropdown still exists and works as intended (🎉) when creating a new commit in the GitHub UI, as described here:
	https://github.com/todogroup/gh-issues/issues/32#issuecomment-406054055

@benbalter , do you have any context or info on why this feature was removed? Thanks for adding it originally!

**Comment by oshaughnessy** [comment]
> The next step would be to allow organization admins to enforce that the corporate email is used for merge commits, which is on our roadmap as a subsequent follow up after the ability to choose the merge commit email.

Out of curiosity, why isn't this the default? It's always seemed logical to me that when I join an org, if I have an email address associated with my GitHub account in that org's verified domains, all my PR approvals and merges should start to use that email address. As things stand now, I'm part of an org with 50+ repos, and as part of the devops team, I'm a reviewer or new committer to new repos all the time. I have to maintain constant vigilance when I'm reviewing and merging to be sure I've chosen the right address. I think GitHub could offer a much more sensible default in this scenario.

**Comment by claudioc** [comment]
Came here because I just merged a PR using my personal email, because I forgot to select the correct email address from the list (it was a new repo for me, hence the primary one was the selected default).

Please add a default to org when using an org repo!

**Comment by eKoopmans** [comment]
Following up on my [earlier comment](https://github.com/todogroup/gh-issues/issues/32#issuecomment-1176759730), I've noticed that there *is* still an author dropdown when choosing `Merge pull request`, but not for `Squash and merge` or `Rebase and merge`.

So my question is: Why have those two been excluded from selecting the author?

**Comment by eKoopmans** [comment]
Not sure if this was announced anywhere, but it looks like the email dropdown is back for squash merges, on both personal and org repos! 🎉 Thanks to whoever was involved in restoring that behaviour!

**Edit:** The change was announced [yesterday on the changelog](https://github.blog/changelog/2022-09-15-git-commit-author-shown-when-squash-merging-a-pull-request/) 😄.

**Comment by tarheel** [comment]
> Not sure if this was announced anywhere, but it looks like the email dropdown is back for squash merges, on both personal and org repos! 🎉 Thanks to whoever was involved in restoring that behaviour!

I came upon this thread because I noticed this change recently (within the last few days, I think) and was hoping to find a way to make the dropdown disappear and just automatically use whichever email I've already chosen for the repo in my `.git/config`. The dropdown does default to the right thing, but I find it distracting that it shows up at all, given that I never want to change it. Apologies if my question is already answered in the long thread above (which I tried to skim), but is there a way to do what I want?

**Comment by eKoopmans** [comment]
> hoping to find a way to make the dropdown disappear

😭 Oh no! It's been such a hassle not having it!

But I suppose on a user end, it would be possible to remove with a script / browser extension - find the element by CSS query and set its display to none. Seems super unlikely that there would be any GitHub-side setting to hide it though.

**Comment by jonnyijapan** [comment]
Clicking in thumb up # 300.

**Comment by a-fortunato** [comment]
> I came upon this thread because I noticed this change recently (within the last few days, I think) and was hoping to find a way to make the dropdown disappear and just automatically use whichever email I've already chosen for the repo in my `.git/config`. 

In the end, I think this comment is asking the same as the rest of the thread: to automate the email it's being used with a configuration (i.e. by repo, by organisation) instead of having a manual setup.

I came to this thread because I too realised that my automated squash-merge (set from the mobile app) and another squash-merge that a colleague did from my PR (web) both include my personal email instead of the professional one. Then every time I squash-merge a PR I have to remember to set/check the correct email (I'm not sure if it's working that feature where the last email used should appear there, but ofc after the two cases mentioned above, the last email used is the personal one, so I can't trust that it will always be the professional one instead).
Honestly, feels a bit chaotic.

I also vote that it would be great to have some kind of automated selection of the preferred email for an organisation, or _at least_ per repo.

Thanks for the great work on the platform by the way :) I don't usually have concerns with it.

**Comment by mattiasJohnson** [comment]
Is the feature to set a commit email per organization when using GH web UI still planned or will it not be added? @benbalter

**Comment by lumaxis** [comment]
I just read through this entire issue to see which gaps remain in GitHub that we could consider tracking as future improvements. GitHub has evolved quite a bit since this issue was originally opened, for example it is now possible to pick the email on squash and rebase PR merges as mentioned in https://github.com/todogroup/gh-issues/issues/32#issuecomment-1249474642 and PR suggestions also respect the default author email.

As described in comments here too, the default author email is chosen by:

1. Looking for the last used email address if the user has previously contributed to the repo
2. ~Alternatively using the email configured in [custom email notification routing](https://docs.github.com/en/account-and-profile/managing-subscriptions-and-notifications-on-github/setting-up-notifications/configuring-notifications#customizing-email-routes-per-organization) on https://github.com/settings/notifications/custom_routing~
3. Falls back to the primary email configured at https://github.com/settings/emails

~Reading for example @a-fortunato's comment, it seems like 1. here might be the most controversial behavior? What are other remaining open gaps that still remain in the product today that you'd like to see addressed?~

Looks like 2. is actually no longer true. It appears that this functionality was removed in 2019 due to the feature causing confusion for some users

**Comment by mattiasJohnson** [comment]
@lumaxis Do you know if 1. always is true for any operations done through the Github Web UI? If so that would pretty much be good enough for me, I would only have to change the primary email once per repo and although not explicit it would mean being able to set an email per repo.

**Comment by gergely-elias** [comment]
I haven't seen this use case mentioned before in this issue, but this seems to affect repositories created from a template owned by an organization.
The author email will be your primary email address even if you have another email set up for that specific organization's notifications.

**Comment by mofojed** [comment]
@lumaxis One part that seems to be missing is setting your primary email address for a given organization or repository. Example:
- We have a private repository, that only specific members/admins can merge changes into, using the "Squash and Merge" strategy
- We have contributors in our organization that have configured their email address in the git repository, and then created a PR
- When the PR is ready to merge, the contributor cannot click the Squash/Merge button, since they do not have permissions. When the admin clicks squash/merge, they do not have the option to select the e-mail address; it automatically uses the contributors primary (personal) email address (which doesn't match the email they set in their git config). 

I would expect to use the email address from their git config, or be able to configure a primary email address specific to that repository/organization. I can't seem to find a way to work around it, other than having the contributor change their primary email address (which they shouldn't have to do).

**Comment by glenpike** [comment]
Please can we also have this for the 'Update Branch' button in PR's - they currently seem to default to my main email address, but I need the merge commit to show the one for my org. 🙏

**Comment by akshaybhalotia** [comment]
A lot has been said already, just here to add a +1

**Comment by DigitalLeaves** [comment]
Adding my 👍

**Comment by xsurfer** [comment]
+1

**Comment by berislavlopac** [comment]
As a bare minimum, I would like to see the option to choose which email address to use for notifications about an organization's repos. E.g. if I'm invited to review a PR on such a repo, I want to be notified at the organization email.

**Comment by disasmwinnie** [comment]
Any feedback from GH whether this is even considered as a feature?

Leaking customer data via notifications to private mail-address is currently a blocker. Have to force people to use a second account.

**Comment by kadazuro** [comment]
hi there, under Settings->Notifications there seems to be a way to do this already, by clicking the custom routing button next to the email.
I was able to setup a custom routing from the repos of an organizations to a specific email (previously registered)
I haven't test it out just yet, but the option is there, fingers crossed.

**Comment by disasmwinnie** [comment]
@kadazuro exactly what I was searching for. Thanks for the tip!

**Comment by berislavlopac** [comment]
Oh you're right @kadazuro , I completely missed that. Thank you!

**Comment by glenpike** [comment]
> hi there, under Settings->Notifications there seems to be a way to do this already, by clicking the custom routing button next to the email. I was able to setup a custom routing from the repos of an organizations to a specific email (previously registered) I haven't test it out just yet, but the option is there, fingers crossed.

This doesn't change the default email for PRs it seems - I set the notification email for my organisation, but for each new Repository I work on, it seems I have to change the email on PRs when merging.

**Comment by dannysauer** [comment]
> Building on that, down the line, we might offer organization admins the ability to restrict which emails can be used for merge commits (e.g., requiring an email in the organization's verified domain be used, and only listing those email in the dropdown). That's a bit further down the line, as we wanted to get the ability to choose your merge commit email out first and in your hands, along with several other improvements, and haven't yet decided if the ability to restrict merge commit emails will be limited to organizations on the Enterprise Cloud plan.

Are we close to being down the line yet, @benbalter? :)  As an org admin for multiple orgs, I'm looking for a way to bring some consistency to our repos (some OSS, some private).  One place that needs work is enforcing that users squash-merge to main (or wherever) with their work email address.  They consistently commit as user@work, thanks to a bot verifying the email addresses, but I haven't found a good way to either make their work email address the default or reject attempts to squash-merge PRs using the wrong email address. I can of course ask everyone to make sure the default is set per repo or make their work email address the primary, but the former is tedious and annoying (make a laceholder commit via the web UI?) and the latter is not really something everyone will buy in to, since the majority of developers also work on private projects, etc.  And either way, I'd have to add a procedure to remind every new hire of the setting, write some auditing scripts, etc.

As a devops person working with a whole bunch of repos, I also would like this feature to make my own work easier. :D

**Comment by dubhunter** [comment]
This is still an issue @benbalter. I created a "Revert" PR through the Web UI, and it had my Primary email. Not my Org email. Can GH just allow per-repo or per-org email settings? This is super frustrating.

**Comment by mikeweilgart** [comment]
So, still not possible to set the default email address to use for PR commit messages on an organization-wide basis?

I'm dealing with dozens and dozens of repos in this organization; having the correct email address only remembered on a per-repo basis is effectively the same as not remembering it at all, since I'll have to maintain constant vigilance anyway to avoid putting my personal email address in the organization's commit history.

I have to agree with [this comment](https://github.com/orgs/community/discussions/4168#discussioncomment-5338168) from community discussions:

> The de facto workaround for this is to have all everyone create a new GitHub account specifically for each organization, which is inconvenient and undermines GitHub's goal of individuals having one account. I'd love to be able to advocate for individuals only having one GitHub account, but until this is resolved I'm inclined to recommend that every company makes employees create a GitHub account specifically for work.

**Comment by XedinUnknown** [comment]
I was able to set up custom routing based on an organization, as explained [here](https://github.com/todogroup/gh-issues/issues/32#issuecomment-1996045274). Whether or not it worked - remains to be determined.

**Comment by ArcanisCz** [comment]
I second this 100%, I am part of 2 organizations and I am breaking some policy constantly by accidentally using wrong email... this is totally business breaking for big companies.

Please note, I have custom routing set - it ONLY AFFECTS NOTIFICATIONS, nothing else.

Also, there are some places where I seem to even cannot be selecting my email.

## Updating branch with main branch
![image](https://github.com/user-attachments/assets/d6433154-835d-4e9e-8bfb-6f7a9835d651)

## Clicking on automerge (merge when ready)
![image](https://github.com/user-attachments/assets/62d4532e-7539-4970-9353-883334e62769)

**Comment by Strernd** [comment]
Would have expected a fix for this after 9 years, but here I am once again, having this problem.

**Comment by skiz** [comment]
Here I am looking around for the expected setting for an org based preferred email and find this is still open years later.

What is the gist of of the last 9 years of progress on this?  It's unprofessional and unexpected behavior to have your personal email be plastered on a merge using the UI.

**Comment by rpggio** [comment]
Looks like I'll have to use 'Keep my email addresses private' for web actions. Then the question is whether to use the same [no-reply address](https://docs.github.com/en/account-and-profile/setting-up-and-managing-your-personal-account-on-github/managing-email-preferences/setting-your-commit-email-address#about-no-reply-email) in git config for sake of consistency. I'm leaning that way - It's not like people won't be able to locate me without an email address.

**Comment by RoyTinker** [comment]
There can only be 1 reason this has not been addressed.

This problem goes away when a company adopts GitHub Enterprise Managed Users (EMU), which is a feature only available to GitHub Enterprise.

It's because supporting this for standard GitHub users would be crossing a feature tier boundary. GitHub wants to upsell your company.

It's unfortunate because it's not the company who pays in the end - it's individual developers. We look dumb when joining a new company that doesn't use GitHub EMU and our first merge commit done through the GitHub UI has a personal email :-(. Then we have to go in and change the setting every time we switch between personal/OSS and work projects -- when clocking in and clocking out of our job.

It's enough to make me consider switching to GitLab for my personal/OSS projects.

**Comment by RoyTinker** [comment]
Another alternative is to create a separate company-specific account for your work, if your company doesn't use GitHub EMU.

**Comment by XedinUnknown** [comment]
If you use a different account, you lose the contributions.

**Comment by lebenitza** [comment]
Hopefully it would not be complicated for GitHub to implement assigning and enforcing (git push / UI actions) using a secondary email address for a specific GitHub Organization you are a part of. So... here's to hoping, it would be amazing to have this.

**Comment by MikeVensel** [comment]
I don't know why anyone is still commenting on this issue. Microsft/Github clearly have no intentions of doing anything about this issue.

**Comment by yangcheng** [comment]
really the issue is still open?

**Comment by helpfulchicken** [comment]
I would love if this was included

**Comment by 678098** [comment]
Same thing. I have my personal email and my work email, I want to sign all my work commits with my work email only. The existing PR suggestions UI doesn't allow this and only shows my personal email with no possibility to change it.
